rattan-core/src/cells/token_bucket.rs

```diff
@@ -0,0 +1,968 @@
+    }
+}
+
+pub struct TokenBucketCellEgress<P>
+where
+    P: Packet,
+{
+    egress: mpsc::UnboundedReceiver<P>,
+    max_size: usize,
+    token_bucket: TokenBucket,
+    peak_token_bucket: TokenBucket,
+    packet_queue: DropTailQueue<P>,
+    next_check: Instant,
+    last_check: Instant,
+    config_rx: mpsc::UnboundedReceiver<TokenBucketCellConfig>,
+    timer: Timer,
+    state: AtomicI32,
+}
+
+impl<P> TokenBucketCellEgress<P>
+where
+    P: Packet + Send + Sync,
+{
+    fn get_new_tb(&self, new_rate: Option<u64>, new_burst: Option<usize>) -> TokenBucket {
+        let mut new_tb = match (new_rate, new_burst) {
+            (None, None) => self.token_bucket,
+            (Some(rate), None) => TokenBucket::new(self.token_bucket.buffer, rate),
+            (None, Some(burst)) => TokenBucket::new(burst, self.token_bucket.rate),
+            (Some(rate), Some(burst)) => TokenBucket::new(burst, rate),
+        };
+        new_tb.tokens = length_to_time(new_tb.buffer, new_tb.rate).as_nanos() as i64;
+        new_tb
+    }
+
+    fn set_config(&mut self, config: TokenBucketCellConfig) {
+        let now = Instant::now();
+        debug!("Previous next_check distance: {:?}", self.next_check - now);
+        debug!("Previous check_point distance: {:?}", now - self.last_check);
+
+        // modify self.token_bucket
+        self.last_check = now;
+
+        self.token_bucket = self.get_new_tb(config.rate, config.burst);
+        self.peak_token_bucket = self.get_new_tb(config.peakrate, config.mtu);
+
+        debug!(
+            "toks: {}, ptoks: {}",
+            self.token_bucket.tokens, self.peak_token_bucket.tokens
+        );
+
+        // calculate max_size
+        self.max_size = min(self.token_bucket.buffer, self.peak_token_bucket.buffer);
+
+        // set queue
+        if let Some(limit) = config.limit {
+            debug!(?limit, "Set inner queue byte_limit: ");
+            self.packet_queue.configure(DropTailQueueConfig::new(
+                None,
+                limit,
+                crate::cells::bandwidth::BwType::NetworkLayer,
+            ));
+        }
+
+        // check if all packets in the queue is smaller than max_size
+        let len = self.packet_queue.length();
+        for _ in 0..len {
+            let front_size = self.packet_queue.get_front_size().unwrap();
+            let packet = self.packet_queue.dequeue().unwrap();
+            if front_size <= self.max_size {
+                self.packet_queue.enqueue(packet);
+            }
+        }
+
+        // check if packet can be delivered
+        self.next_check = now;
+    }
+
+    // calculate and return next_check Instant
+    fn calculate_next_check(&self, now: Instant, toks: i64, ptoks: i64) -> Instant {
+        let tokens = min(toks, ptoks);
+        if tokens >= 0 {
+            max(now, self.next_check)
+        } else {
+            let wait_duration = Duration::from_nanos((-tokens) as u64);
+            now + wait_duration
+        }
+    }
```

Comment:
I really dont know how to sleep the cell until next opportunity to send a packet in tbf.

Acturally in `Bandwidth` cell we can use `next_available` to mark the next point to send a packet, because the time point to send the next packet depends on the size of the **current** packet.

But in `TokenBucket` cell, the time point to send the next packet depends on the size of the **next** packet. When I try to send a packet, I have to check the packet in the front of the queue (that's the next packet), check if tokens are enough to send it, and set `next_check` to the time point it can be sent if tokens are not enough.

Maybe you ask why we dont check the next packet when we send a packet? Then how to deal with the situation where there is only one packet in the queue? We have to wait for a new packet and check if tokens are enough to send the new packet and set `next_check` to the time point it can be sent if tokens are not enough.

Reply:
In TBF, we can still use `next_available`, as permitted by its internal state machine: we wait only for configuration information and packets until the queue exactly contains packets (to break the first loop). The conditions are: 1. when the queue transits from empty to having packets, we break the loop; 2. when the `next_available` approaches `now`, we break the loop. Only after that do we have to check if tokens are enough to transmit packets. If yes, we should check for and modify `next_available` before returning `Some(packet)` (set to the next possible timing if there are still packets in the queue or infinitely large if there is no packet).

Reply:
> If yes, we should check for and modify next_available before returning Some(packet) (set to the next possible timing if there are still packets in the queue or infinitely large if there is no packet).

What should I do when tokens are not enough to transmit the packet?

Reply:
> > If yes, we should check for and modify next_available before returning Some(packet) (set to the next possible timing if there are still packets in the queue or infinitely large if there is no packet).
> 
> What should I do when tokens are not enough to transmit the packet?

Calculate the possible `Instant` the packet could be sent and set the `next_available` to it.